### PR TITLE
[Implement] batch change ownership; fixes #640

### LINKED ISSUE #640: As a user i want to change the ownership of an ElementDefinition, including the ownership of its contained Parameters
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the CDP4-IME
- [x] I have searched [open](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
As a user i want to change the ownership of an ElementDefinition, including the ownership of its contained Parameters. This shall be achieved by adding a context menu item (right mouse click) on an ElementDefinition. A dialog shall be popped to select the domain of expertise.

### Steps to Reproduce
NA

### Logs

  - [x] There are no logs regarding the issue
  
### System Configuration

- CDP4 version:
  - [x] CDP4-IME: 7.0.0.4

### PR DESCRIPTION
**Pull request by samatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
[Add] ChangeOwnerShipBatchService to update multiple Things in a batch operation, applied to ElementDefnintionBrowser to change ownership of a selected ElementDefinition and the contained Parameters

### CODE REVIEW

File: EngineeringModel/EngineeringModelModule.cs [hunk 1]
```diff
@@ -129,6 +133,11 @@ public EngineeringModelModule(IRegionManager regionManager, IFluentRibbonManager
         /// </summary>
         internal IParameterActualFiniteStateListApplicationBatchService ParameterActualFiniteStateListApplicationBatchService { get; private set; }
 
+        /// <summary>
+        /// The <see cref="IChangeOwnerShipBatchService"/> used to change the ownership of multiple <see cref="IOwnedThing"/>s in a batch operation
+        /// </summary>
+        internal IChangeOwnerShipBatchService ChangeOwnerShipBatchService { get; private set; }
```
Comment (alexatrhea) [inline EngineeringModel/EngineeringModelModule.cs:139]: The "S" should not be capital in "Ownership" is the case of class, interface and variable names.

File: EngineeringModel/Services/ChangeOwnerShipBatchService.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    using CDP4Dal;
+    using CDP4Dal.Operations;
+
+    /// <summary>
+    /// The purpose of the <see cref="ChangeOwnerShipBatchService"/> is to change the ownership of multiple
+    /// <see cref="IOwnedThing"/>s in a batch operation
+    /// </summary>
+    [Export(typeof(IChangeOwnerShipBatchService))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class ChangeOwnerShipBatchService : IChangeOwnerShipBatchService
```
Comment (alexatrhea) [inline EngineeringModel/Services/ChangeOwnerShipBatchService.cs:47]: The name of the class, same as comment before.

File: EngineeringModel/Services/ChangeOwnerShipBatchService.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    {
+        /// <summary>
+        /// Update the ownnership of ultiple <see cref="IOwnedThing"/>s
```
Comment (alexatrhea) [inline EngineeringModel/Services/ChangeOwnerShipBatchService.cs:50]: spelling in "multiple"

File: EngineeringModel/ViewModels/Dialogs/ChangeOwnerShipSelectionDialogViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    using ReactiveUI;
+
+    /// <summary>
+    /// The purpose of the <see cref="ChangeOwnerShipSelectionDialogViewModel"/> is to facilitate the selection
+    /// of the new owner for an <see cref="IOwnedThing"/> and it's contained <see cref="IOwnedThing"/>
+    /// </summary>
+    /// <remarks>
+    /// this allows a user to change the ownership of an <see cref="ElementDefinition"/> and it's contained <see cref="Parameter"/>s
+    /// </remarks>
+    public class ChangeOwnerShipSelectionDialogViewModel : DialogViewModelBase
```
Comment (alexatrhea) [inline EngineeringModel/ViewModels/Dialogs/ChangeOwnerShipSelectionDialogViewModel.cs:46]: Ship should not be with capital letter

File: EngineeringModel/ViewModels/Dialogs/ChangeOwnerShipSelectionDialogViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    {
+        /// <summary>
+        /// Backing field for the <see cref="SelectedOwner"/> property.
+        /// </summary>
+        private DomainOfExpertise selectedOwner;
+
+        /// <summary>
+        /// Backing field for the <see cref="IsContainedItemChangeOwnerShipSelected"/> property.
+        /// </summary>
+        private bool isContainedItemChangeOwnerShipSelected;
```
Comment (alexatrhea) [inline EngineeringModel/ViewModels/Dialogs/ChangeOwnerShipSelectionDialogViewModel.cs:56]: Ship should not be with capital letter

File: EngineeringModel/ViewModels/Dialogs/ChangeOwnerShipSelectionResult.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    using CDP4Common.SiteDirectoryData;
+
+    using CDP4Composition.Navigation;
+
+    /// <summary>
+    /// The purpose of the <see cref="ChangeOwnerShipSelectionResult"/> is to present possible <see cref="DomainOfExpertise"/>s and to
+    /// to select one. A selection that allows the user to indicate whether the contained items shall have their ownership changed as
+    /// well is also present
+    /// </summary>
+    public class ChangeOwnerShipSelectionResult : BaseDialogResult
```
Comment (alexatrhea) [inline EngineeringModel/ViewModels/Dialogs/ChangeOwnerShipSelectionResult.cs:37]: Ship should not be with capital letter

File: EngineeringModel/ViewModels/Dialogs/ChangeOwnerShipSelectionResult.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        /// <summary>
+        /// Gets or sets the selected <see cref="DomainOfExpertise"/>
+        /// </summary>
+        public DomainOfExpertise DomainOfExpertise { get; private set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the contained <see cref="IOwnedThing"/>s shall be have the
+        /// owner property updated as well.
+        /// </summary>
+        public bool IsContainedItemChangeOwnerShipSelected { get; private set; }
```
Comment (alexatrhea) [inline EngineeringModel/ViewModels/Dialogs/ChangeOwnerShipSelectionResult.cs:58]: Ship should not be with capital letter

File: EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionsBrowserViewModel.cs [hunk 8]
```diff
@@ -510,6 +528,8 @@ public override void PopulateContextMenu()
                 this.ContextMenu.Insert(1, new ContextMenuItemViewModel("Create a Parameter Group", "", this.CreateParameterGroup, MenuItemKind.Create, ClassKind.ParameterGroup));
                 this.ContextMenu.Insert(2, new ContextMenuItemViewModel("Copy the Element Definition", "", this.CopyElementDefinitionCommand, MenuItemKind.Copy, ClassKind.ElementDefinition));
                 this.ContextMenu.Insert(3, new ContextMenuItemViewModel("Highlight Element Usages", "", this.HighlightElementUsagesCommand, MenuItemKind.Highlight, ClassKind.ElementUsage));
+                this.ContextMenu.Insert(4, new ContextMenuItemViewModel("Change OwnerShip", "", this.ChangeOwnerShipCommand , MenuItemKind.Edit, ClassKind.NotThing));
```
Comment (alexatrhea) [inline EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionsBrowserViewModel.cs:531]: Ship should not be with capital letter

File: EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionsBrowserViewModel.cs [hunk 9]
```diff
@@ -914,6 +934,46 @@ private async Task ExecuteBatchCreateSubscriptionCommand()
             }
         }
 
+        /// <summary>
+        /// Execute the <see cref="ChangeOwnerShipCommand"/>
+        /// </summary>
+        /// <returns>
+        /// an awaitable <see cref="Task"/>
+        /// </returns>
+        private async Task ExecuteChangeOwnerShipCommand()
```
Comment (alexatrhea) [inline EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionsBrowserViewModel.cs:943]: Ship should not be with capital letter

File: EngineeringModel/Views/Dialogs/ChangeOwnerShipSelectionDialog.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    using CDP4Composition.Attributes;
+    using CDP4Composition.Navigation.Interfaces;
+
+    using DevExpress.Xpf.Core;
+
+    /// <summary>
+    /// Interaction logic for ChangeOwnerShipSelectionDialog.xaml
+    /// </summary>
+    [DialogViewExport("ChangeOwnerShipSelectionDialog", "Ownership Selection")]
+    public partial class ChangeOwnerShipSelectionDialog : DXWindow, IDialogView
```
Comment (alexatrhea) [inline EngineeringModel/Views/Dialogs/ChangeOwnerShipSelectionDialog.xaml.cs:37]: Ship should not be with capital letter

**Review by alexatrhea — CHANGES_REQUESTED** [review]
JUst a cosmetic change

**Review by nathanatrhea — APPROVED** [review]
no other comment than Alex's comment

**Review by ahmedatrhea — APPROVED** [review]

**Pull request closed without merging**